src/python/gudhi/wasserstein/wasserstein.py

```diff
@@ -88,32 +93,150 @@ def _perstot(X, order, internal_p, enable_autodiff):
+    :returns: cost and matching due to essential parts. If cost is +inf, matching will be set to None.
+    '''
+    ess_parts_X = _get_essential_parts(X)
+    ess_parts_Y = _get_essential_parts(Y)
+
+    # Treats the case of infinite cost (cardinalities of essential parts differ).
+    for u, v in zip(ess_parts_X, ess_parts_Y):
+        if len(u) != len(v):
+            return np.inf, None
+
+    # Now we know each essential part has the same number of points in both diagrams.
+    # Handle type 0 and type 1 essential parts (those with one finite coordinates)
+    c1, m1 = _cost_and_match_essential_parts(X, Y, ess_parts_X[0], ess_parts_Y[0], axis=1, order=order)
+    c2, m2 = _cost_and_match_essential_parts(X, Y, ess_parts_X[1], ess_parts_Y[1], axis=0, order=order)
+
+    c = c1 + c2
+    m = m1 + m2
+
+    # Handle type >= 2 (both coordinates are infinite, so we essentially just align points)
+    for u, v in zip(ess_parts_X[2:], ess_parts_Y[2:]):
+        m += list(zip(u, v))  # cost is 0
+
+    return c, np.array(m)
+
+
+def _finite_part(X):
+    '''
+    :param X: (n x 2) numpy array encoding a persistence diagram.
+    :returns: The finite part of a diagram `X` (points with finite coordinates).
+    '''
+    return X[np.where(np.isfinite(X[:,0]) & np.isfinite(X[:,1]))]
+
+
+def _warn_infty(matching):
+    '''
+    Handle essential parts with different cardinalities. Warn the user about cost being infinite and (if
+    `matching=True`) about the returned matching being `None`.
+    '''
+    if matching:
+        warnings.warn('Cardinality of essential parts differs. Distance (cost) is +infty, and the returned matching is None.')
+        return np.inf, None
+    else:
+        warnings.warn('Cardinality of essential parts diffes. Distance (cost) is +infty.')
+        return np.inf
+
+
+
+def wasserstein_distance(X, Y, matching=False, order=1., internal_p=np.inf, enable_autodiff=False,
+                         keep_essential_parts=True):
     '''
-    :param X: (n x 2) numpy.array encoding the (finite points of the) first diagram. Must not contain essential points
-                (i.e. with infinite coordinate).
+    :param X: (n x 2) numpy.array encoding the first diagram. Can contain essential parts (points with infinite
+                        coordinates).
     :param Y: (m x 2) numpy.array encoding the second diagram.
     :param matching: if True, computes and returns the optimal matching between X and Y, encoded as
                      a (n x 2) np.array  [...[i,j]...], meaning the i-th point in X is matched to
                      the j-th point in Y, with the convention (-1) represents the diagonal.
-    :param order: exponent for Wasserstein; Default value is 1.
-    :param internal_p: Ground metric on the (upper-half) plane (i.e. norm L^p in R^2);
+                     Note that if the cost is +inf (essential parts have different number of points,
+                     then the optimal matching will be set to `None`.
```

Comment:
We could also return an arbitrary matching, say everyone to the diagonal, if that helps callers have a more uniform behavior. But it is probably better indeed if callers are forced to check for inf/None (as in this PR) and don't try to use an arbitrary matching.

Reply:
I don't know what is best. 
Returning an arbitrary matching may lead to quite unexpected behavior when it comes to debug as the code is likely to run flawless; thus I personally prefer returning `None` so that the code is likely to crash and the user will first have to figure out how to handle this (e.g. setting `keep_essential_parts=False`).